[POS][Local Catalog] Connect POS to WooPosVariationEntity table

samiuelson opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the Linear issue this PR addresses, e.g., WOOMOB-373. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
The goal of this PR is to load variations from the DB instead of the existing in-memory cache / remote call hybrid store. Currently, the logic that decides on which data source to use is based on the feature flag `WOO_POS_LOCAL_CATALOG_M1` state.

💡 PTR is not working in variation list. Incremental sync on PTR will be added in a separate PR (WOOMOB-1069).

### Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->
—

### Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->
1. Verify that in the release build (FF disabled), the previous data source implementation is used.
2. Verify that PTR works when FF is disabled.
3. When the FF is enabled, verify that variations load correctly. They should load significantly faster.

### The tests that have been performed
<!-- To give the reviewer an idea of what could be missed in terms of testing -->
Above

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->
No visual changes are expected.

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

samiuelson committed:
Load variations from DB in case FF is enabled

samiuelson committed:
Load variations from DB in case FF is enabled

samiuelson committed:
Merge branch 'woomob-1244-woo-poslocal-catalog-use-new-pos-specific-variation-model-in' into woomob-1073-woo-poslocal-catalog-connect-variations-to-pos-variations

# Conflicts:
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/variations/WooPosVariationsDataSource.kt
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/variations/WooPosVariationsViewModel.kt
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/localcatalog/WooPosSyncProductsAction.kt

dangermattic commented:
<!--
  1 Error: Please add tests for class <co...
  1 Warning: This PR is larger than 300 lin...
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please add tests for class <code>WooPosVariationsInDbDataSource</code> (or add <code>unit-tests-exemption</code> label to ignore this).</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

samiuelson committed:
Merge branch 'trunk' into woomob-1073-woo-poslocal-catalog-connect-variations-to-pos-variations

# Conflicts:
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/variations/WooPosVariationsDataSource.kt
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/variations/WooPosVariationsViewModel.kt

samiuelson committed:
Merge remote-tracking branch 'origin/woomob-1073-woo-poslocal-catalog-connect-variations-to-pos-variations' into woomob-1073-woo-poslocal-catalog-connect-variations-to-pos-variations

# Conflicts:
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/variations/WooPosVariationsDataSource.kt
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/variations/WooPosVariationsInDbDataSource.kt
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/variations/WooPosVariationsViewModel.kt
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/localcatalog/WooPosSyncProductsAction.kt

samiuelson committed:
Merge remote-tracking branch 'origin/woomob-1073-woo-poslocal-catalog-connect-variations-to-pos-variations' into woomob-1073-woo-poslocal-catalog-connect-variations-to-pos-variations

# Conflicts:
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/variations/WooPosVariationsDataSource.kt
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/variations/WooPosVariationsInDbDataSource.kt
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/variations/WooPosVariationsViewModel.kt
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/localcatalog/WooPosSyncProductsAction.kt

samiuelson committed:
Merge remote-tracking branch 'origin/woomob-1073-woo-poslocal-catalog-connect-variations-to-pos-variations' into woomob-1073-woo-poslocal-catalog-connect-variations-to-pos-variations

samiuelson committed:
Clean up code

samiuelson committed:
Fix variation result mapping

samiuelson committed:
Clean up code - remove comments

samiuelson committed:
Update comment

samiuelson committed:
Clean up code

samiuelson committed:
Fix WooPosVariationsViewModelTest after merge

- Added missing dependencies: WooPosVariationsInDbDataSource and WooPosLocalCatalogM1Enabled
- Fixed constructor parameter order to match the actual ViewModel
- Added missing import for doReturn from mockito-kotlin
- All variation-related tests now pass

samiuelson committed:
Merge branch 'trunk' into woomob-1073-woo-poslocal-catalog-connect-variations-to-pos-variations

samiuelson committed:
Satisfy detekt's complaints

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/variations/WooPosVariationsInDbDataSource.kt:
```diff
@@ -0,0 +1,55 @@
+            .map { result ->
+                result.getOrNull()?.map { it.toWooPosVariation(mapper) } ?: emptyList()
+            }
+            .firstOrNull() ?: emptyList()
```
I believed the plan was to continuously subscribe to the database and display all the modifications in the table, correct?

samiuelson replied:
Good catch, fixed: 080d9a3

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/variations/WooPosVariationsViewModel.kt:
```diff
@@ -34,12 +35,20 @@ class WooPosVariationsViewModel @Inject constructor(
 
+    private val currentDataSource: WooPosVariationsDataSourceInterface
+        get() = when (wooPosLocalCatalogM1Enabled()) {
+            true -> variationsInDbDataSource
```
np: you call `wooPosLocalCatalogM1Enabled` everytime when access `currentDataSource`. Maybe to use field not property then?

samiuelson replied:
Done: c007629

kidinov commented:
Not sure if this is expected, but It crashes now when a variation is selected on db data source

```
Process: com.woocommerce.android.dev, PID: 25840
                                                                                                    java.lang.NullPointerException
                                                                                                    	at com.woocommerce.android.ui.woopos.home.cart.WooPosCartViewModel.handleVariationClicked(WooPosCartViewModel.kt:344)
                                                                                                    	at com.woocommerce.android.ui.woopos.home.cart.WooPosCartViewModel.access$handleVariationClicked(WooPosCartViewModel.kt:56)
                                                                                                    	at com.woocommerce.android.ui.woopos.home.cart.WooPosCartViewModel$handleVariationClicked$1.invokeSuspend(Unknown Source:18)
                                                                                                    	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
                                                                                                    	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:100)
                                                                                                    	at android.os.Handler.handleCallback(Handler.java:942)
                                                                                                    	at android.os.Handler.dispatchMessage(Handler.java:99)
                                                                                                    	at android.os.Looper.loopOnce(Looper.java:201)
                                                                                                    	at android.os.Looper.loop(Looper.java:288)
                                                                                                    	at android.app.ActivityThread.main(ActivityThread.java:7924)
                                                                                                    	at java.lang.reflect.Method.invoke(Native Method)
                                                                                                    	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
                                                                                                    	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:936)
                                                                                                    	Suppressed: kotlinx.coroutines.internal.DiagnosticCoroutineContextException: [StandaloneCoroutine{Cancelling}@58df775, Dispatchers.Main.immediate]
```

kidinov reviewed (approved):
LGTM! I don't know what the expectations are, so I left a few comments, please take a look

samiuelson committed:
Fix crash when selecting a variation

samiuelson committed:
Make WooPosVariationsInDbDataSource observable

samiuelson committed:
Declare field instead of a property

samiuelson committed:
Satisfy detekt's complaints

samiuelson committed:
Clean up code

samiuelson commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/variations/WooPosVariationsInDbDataSource.kt:
```diff
@@ -0,0 +1,54 @@
+    private fun getVariationsFromDatabaseFlow(productId: Long): Flow<List<WooPosVariation>> {
+        val siteModel = selectedSite.getOrNull() ?: return flow { emit(emptyList()) }
+        val siteId = LocalId(siteModel.id)
+        val remoteProductId = RemoteId(productId)
```
Fixed: 223a3c3

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/cart/WooPosCartViewModel.kt:
```diff
@@ -341,7 +347,16 @@ class WooPosCartViewModel @Inject constructor(
         val itemNumber = getItemNumber()
-        val productVariation = getVariationsById(productId, variationId)!!
+
+        val productVariation = when {
```
Have you considered to move this logic into `WooPosGetVariationById`? Imo it would be cleaner than keeping this in the VMs

samiuelson replied:
I think I'll leave the FF conditional switch in the VM for now so that it's easier to track, and we will remove it soon anyway.

samiuelson committed:
Fix tests

samiuelson committed:
Satisfy detekt's complaints

samiuelson committed:
Merge branch 'trunk' into woomob-1073-woo-poslocal-catalog-connect-variations-to-pos-variations

samiuelson committed:
Fix tests
